main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
+        if (i < 0)
+            s
+        else {
+            let s1 = f(a[i], s);
+            foldRightHelper(f, s1, a, i - 1)
+        }
+
+    ///
+    /// Alias for `reduceLeft`.
+    ///
+    pub def reduce(f: (a, a) -> a & e, a: Array[a]): Option[a] & e = reduceLeft(f, a)
+
+    ///
+    /// Applies `f` to all elements in `a` going from left to right until a single value `v` is obtained.  Returns `Some(v)`.
+    ///
+    /// Returns `None` if `a` is empty.
+    ///
+    pub def reduceLeft(f: (a, a) ->  a & e, a: Array[a]): Option[a] & e =
+        if (length(a) == 0)
+            None
+        else
+            Some(reduceLeftHelper(f, a, a[0], 1)) as & e
+
+    ///
+    /// Helper function for `reduceLeft`.
+    ///
+    def reduceLeftHelper(f: (a, a) ->  a & e, a: Array[a], x: a, i:Int32): a & Impure =
+        if (i >= length(a))
+            x
+        else
+            reduceLeftHelper(f, a, f(x, a[i]), i+1)
+
+    ///
+    /// Applies `f` to all elements in `a` going from right to left until a single value `v` is obtained.  Returns `Some(v)`.
+    ///
+    /// Returns `None` if `a` is empty.
+    ///
+    pub def reduceRight(f: (a, a) -> a & e, a: Array[a]): Option[a] & e =
+        if (length(a) == 0)
+            None
+        else {
+            let end = length(a) - 1;
+            let ac = a[end] as & e;
+            Some(reduceRightHelper(f, a, ac, end-1)) as & e
+        }
+
+    ///
+    /// Helper function for `reduceRight`.
+    ///
+    def reduceRightHelper(f: (a, a) -> a & e, a: Array[a], x: a, i:Int32): a & Impure =
+        if (i < 0)
+            x
+        else
+            reduceRightHelper(f, a, f(a[i], x), i-1)
+
+    ///
+    /// Returns the number of elements in `a` that satisfy the predicate `f`.
+    ///
+    pub def count(f: a -> Bool, a: Array[a]): Int32 =
+        foldLeft((b,x) -> if (f(x)) b + 1 else b, 0, a)
+
+    ///
+    /// Returns the concatenation of the arrays of in the array `arrs`.
+    ///
+    pub def flatten(arrs: Array[Array[a]]) : Array[a] & Impure =
+        let len = sumLengths(arrs);
+        match headArrays(arrs) {
+            case None => []
+            case Some(x) => {
+                let out = Array.new(x, len);
+                foreachAccum((a, pos) -> arrayWrites(pos, a, out), 0, arrs);
+                out
+            }
+        }
+
+    ///
+    /// Returns `true` if and only if at least one element in `a` satisfies the predicate `f`.
+    ///
+    /// Returns `false` if `a` is empty.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def exists(f: a -> Bool, a: Array[a]): Bool =
+        existsHelper(f, a, length(a), 0) as & Pure
+
+    ///
+    /// Helper function for `exists`.
+    ///
+    def existsHelper(f: a -> Bool, a: Array[a], len: Int32, i: Int32): Bool & Impure =
+        if (i >= len)
+            false
+        else
+            if (f(a[i]) == true)
+                true
+            else
+                existsHelper(f, a, len, i+1)
```

Comment:
I like this style of formatting (absent any switch feature). Maybe it could be used on line 143 too?